### PR TITLE
chore: prepare sessions for encryption (AR-971)

### PR DESCRIPTION
**Pull request by vitorhugods**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] contains a reference JIRA issue number like `SQPIT-764`
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

When sending a message, we need to make sure we have cryptographic session with all the recipients. If we don't, then we need to fetch PreKeys for each missing session and establish them.

### Solutions

Created a `OutgoingMessageRecipientsRetriever` which needs a better name.

### Dependencies

Needs `PreKeyRepository`, which is not in place yet. 

* #223 

### Testing

[W.I.P]

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: cryptography/src/jvmMain/kotlin/com/wire/kalium/cryptography/ProteusClientImpl.kt [hunk 1]
```diff
@@ -45,6 +45,10 @@ actual class ProteusClientImpl actual constructor(rootDir: String, userId: Strin
         return wrapException { box.newPreKeys(from, count).map { toPreKey(it) } as ArrayList<PreKey> }
     }
 
+    override suspend fun doesSessionExist(sessionId: CryptoSessionId): Boolean {
+        TODO("Nope. Can't do this using cryptobox4j.")
```
Comment (vitorhugods) [inline cryptography/src/jvmMain/kotlin/com/wire/kalium/cryptography/ProteusClientImpl.kt:49]: I think we can discuss with @LukasForst and see if we can turn [CryptoBox#getSession](https://github.com/wireapp/cryptobox4j/blob/4af990ad1874f08c42d306e0b2d0e7646f72f027/src/main/java/com/wire/bots/cryptobox/CryptoBox.java#L334) public and publish a new version of Cryptobox4j.

Comment (LukasForst) [reply cryptography/src/jvmMain/kotlin/com/wire/kalium/cryptography/ProteusClientImpl.kt:49]: Sure, I can do that 👍

Or I can add function for checking if the session exists (preferred), the way I see it, Cryptobox4j was designed in a way that consumers of its API won't have access to session directly. Is that enough for you @vitorhugods? Something like `Cryptobox.doesSessionExist(sessionId: String): Boolean`?

Comment (LukasForst) [reply cryptography/src/jvmMain/kotlin/com/wire/kalium/cryptography/ProteusClientImpl.kt:49]: `doesSessionExist` is now in release `1.1.1` - https://github.com/wireapp/cryptobox4j/blob/master/src/main/java/com/wire/bots/cryptobox/CryptoBox.java#L347

Comment (vitorhugods) [reply cryptography/src/jvmMain/kotlin/com/wire/kalium/cryptography/ProteusClientImpl.kt:49]: @LukasForst I just went for a nap and you solved it already 😍
This is enough, yes! Thank you very much.

**Review by typfel — APPROVED** [review]

**Review by MohamadJaara — APPROVED** [review]

**Pull request merged**